### ISSUE TITLE
[FTBFS] cargo for x86_64-musl

### ISSUE THREAD

**Issue by pullmoll** [void-linux/void-packages#12058]
```
…
   Compiling jobserver v0.1.13
error[E0308]: mismatched types
   --> /tmp/.cargo/registry/src/github.com-1ecc6299db9ec823/jobserver-0.1.13/src/lib.rs:665:40
    |
665 |                     libc::pthread_kill(self.thread.as_pthread_t(), libc::SIGUSR1);
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^ expected *-ptr, found u64
    |
    = note: expected type `*mut imp::libc::c_void`
               found type `u64`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: Could not compile `jobserver`.
warning: build failed, waiting for other jobs to finish...
error: build failed
=> ERROR: cargo-0.36.0_1: do_build: '$cargo build --release $(vopt_if static --features="all-static")' exited with 101
=> ERROR:   in do_build() at srcpkgs/cargo/template:103
```

**Comment by pullmoll** [comment]
Now trying to build rust for x86_64-musl with this patch:
```
--- rustc-1.35.0-src/vendor/libc/src/unix/notbsd/linux/mod.rs   2019-05-20 15:47:24.000000000 +0200
+++ rustc-1.35.0-src/vendor/libc/src/unix/notbsd/linux/mod.rs   2019-06-01 12:57:27.526582596 +0200
@@ -3,7 +3,7 @@
 pub type useconds_t = u32;
 pub type dev_t = u64;
 pub type socklen_t = u32;
-pub type pthread_t = c_ulong;
+pub type pthread_t = u64;
 pub type mode_t = u32;
 pub type ino64_t = u64;
 pub type off64_t = i64;
```
and see if that fixes building cargo.
I have no experience with rust and cargo so this patch may be plain stupid.

**Comment by jnbr** [comment]
Interesting. Cargo doesn't ship with a `Caro.lock` file, so the versions of the dependencies can change and the builds aren't reproducible. The problem is most likely because of a change in the rust `libc` crate between version 0.2.56 and 0.2.57.

**Comment by jnbr** [comment]
here is a workaround to use a fixed (working) version of libc:

```
--- a/srcpkgs/cargo/template
+++ b/srcpkgs/cargo/template
@@ -100,6 +100,9 @@ do_build() {
                cargo="./cargo"
        fi

+       $cargo update
+       $cargo update --package libc --precise 0.2.55
+
        $cargo build --release $(vopt_if static --features="all-static")
 }
```

**Comment by pullmoll** [comment]
Thanks @jnbr this indeed fixes the cargo build for x86_64-musl.

**Comment by jnbr** [comment]
The definition of `libc::pthread_t` was changed [here](https://github.com/rust-lang/libc/commit/8f403e5ac262508b763da173d545195710d10494#diff-f78d94e8743c5dfc443c2b1351137950R1)

There is a PR for rustc to handle the breakage: https://github.com/rust-lang/rust/pull/61428

**Comment by jnbr** [comment]
Fixed for now with b68db8e7f60a6e5ce57e77d7f6865b9bc7b28f76
